File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys 
input = sys.stdin.readline

n, c = map(int, input().split())
arr = []
for i in range(n):
    arr.append(int(input()))
arr.sort()
start, end = 1, arr[-1]-arr[0]
answer = 0

while start <= end:
    mid = (start+end)//2
    curr = arr[0]
    cnt = 1
    
    for i in range(1, len(arr)):
        if arr[i] >= curr+mid:
            cnt+=1
            curr = arr[i]
            
    if cnt >= c:
        start = mid+1
        answer = start-1
    else:
        end = mid-1
        
print(answer)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int N = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());
        int[] arr = new int[N];
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(arr);

        int ans = 0;
        int low = 0;
        int high = arr[N-1]-arr[0];
        while (low <= high) {
            int cnt = 1;
            int cur = arr[0];
            int mid = (low+high)/2;
            
            for (int i = 1; i < N; i++) {
                if (arr[i]-cur>=mid) {
                    cnt++;
                    cur = arr[i];
                }
            }
            if (cnt >= C) {
                ans = mid;
                low = mid+1;
            } else {
                high = mid-1;
            }
        }
        System.out.println(ans);
    }
}
